typing/ctype.ml

```diff
@@ -5328,7 +5374,8 @@ let rec normalize_type_rec mark ty =
         begin match !nm with
         | None -> ()
         | Some (n, v :: l) ->
-            if deep_occur ty (newgenty (Ttuple l)) then
+            if deep_occur ty (newgenty (Ttuple (List.map (fun t -> None, t) l)))
```

Comment:
Previously `Ttuple` was used as a hack to run `deep_occur` on a list of types. This should be equivalent to `List.exists (deep_occur ty) l`, but it will be faster if some of the elements of the list are shared or share subtypes -- they benefit from a single visited-map and will not be traversed several times.

This was inelegant before and I find it slightly worse now. I would prefer if we had a `deep_occur_list` version that checks for a list of types. This could possibly be implemented with something like:

```ocaml
type 'a inputs = One of 'a | Several of 'a list
val deep_occur_gen : type -> type inputs -> bool

(* easy wrappers over deep_occur_gen *)
val deep_occur
val deep_occur_list
```

Reply:
Yes, this is true.  @rtjoa actually pointed out the same thing when he was working on this in our branch, and we merged a fix for `deep_occur` (ocaml-flambda/flambda-backend#1503) and he also wrote one for `free_vars` (ocaml-flambda/flambda-backend#1519) but we didn't merge it because it costs us something to have diffs with the main OCaml compiler.

I deleted his diff for `deep_occur` when backporting this just because `deep_occur` had changed a bit upstream and I wanted to focus on the core functionality rather than generalizing the new type marking mechanism.  But I agree that what you and Ryan have suggested is cleaner and I will make an additional commit here with those changes in the next day or two.

Reply:
I've now pushed a commit that creates list versions of `deep_occur` and `free_vars`. I agree this is a nice improvement. I didn't find it necessary to add a type like `inputs` - let me know what you think.